### ISSUE TITLE
Injecting CSS Compiler

### ISSUE THREAD

**Issue by p3k** [ractivejs/ractive#3242]
(This might be a follow-up to #3044.)

We are currently using Ractive and LESS together with RequireJS. Our build workflow compiles all LESS files to one CSS file, the Ractive code and some components are becoming a scripts bundle.

We now would like to use Ractive components generally, making use of their scoped CSS feature, but still be able to use LESS syntax.

However, so far I did not succed to combine the LESS compiler with the Ractive CSS.

Either I get the compiled CSS directly injected in the page using the requirejs-less plugin, but no CSS scoping from Ractive…

Or I get the scoped CSS but cannot use LESS with it.

Does anyone have a suggestion how to tackle this problem? Is there a hook in Ractive I could use to attach the LESS parser? Or some other idea that might be worthwhile to try?

**Comment by evs-chris** [comment]
Are you pre-compiling your components? Either way, `css` is just a string, so you can process the less separately and pull it in via a requirejs plugin or in your build process and just assign it to the css extend option e.g. (and forgive me, I've never used requirejs personally, but hopefully this conveys the intent)

```js
define('mycomponent', ['ractive', 'some/style.less'], function(Ractive, css) {
  return Ractive.extend({
    template: `...`,
    css
  });
}
```

If you're using ractive single file components via ractive-load or something like it, you can usually `<link>` in stylesheets, so you can have that file preprocessed as well.

**Comment by ghost** [comment]
Are you using component files or separate JS and LESS?

If it's the former (component files), Ractive does have a RequireJS plugin called [rvc](https://github.com/ractivejs/rvc) but I don't remember it supporting pre-processing the contents of `<style>`. You can probably fork rvc and tweak how it builds the constructor to insert LESS compilation. Under the hood, it uses [rcu](https://github.com/ractivejs/rcu) to parse a component file into its constituent parts. Then rvc then either constructs a constructor (runtime) or compiles code (build time). 

If it's the latter (separate JS and LESS), I'd look for a LESS RequireJS plugin and do @evs-chris 's suggestion. It's probably lower effort than extending/forking rvc to allow CSS pre-processing.

**Comment by p3k** [comment]
Thanks for the suggestions @evs-chris and @fskreuz.

We are currently using both, component files as well as separate JS and LESS but so far I only managed to process the LESS code with the latter. (Component files also do not seem to allow to transpile their `<script>` code, so I cannot use ES6 there.)

Here is the blueprint of what I tried (I am using the ES6 plugin for RequireJS should you get confused about no `require()` statements).

First, the approach using the less plugin for RequireJS:

```js
import Foo from 'es6!components/foo/Foo';
import template from 'text!./Component.html';
import css from 'less!./Component.less';

const Component = Ractive.extend({
  components: { Foo },
  template,
  css
});

export { Component };
```

This perfectly compiles the LESS code to CSS – but the result is injected as `<style>` element into the page, the  `css` variable (and thus, the `css` property of the `extend()` argument) actually is `undefined`!

So I get my CSS 😺 but I don’t get it scoped to the component template :slightly_frowning_face: 

So I tried this variant: 

```js
import Foo from 'es6!components/foo/Foo';
import template from 'text!./Component.html';
import less from 'text!./Component.less';
import lessc from 'lessc';

let css = '';

lessc.render(less, {}, (error, result) => {
   if (result) // …?
});

const Component = Ractive.extend({
  components: { Foo },
  template,
  css
});

export { Component };
```

Here, I load the LESS code as is using the RequireJS text plugin as well as the LESS compiler `lessc` and then call its `render()` method – just to get stuck with the asynchronous code.

So it could be my actual issue is due to the asynchronous nature of the LESS compiler – so far I have not found a way to let it run synchronously, although I tried the suggestions in less/less#3017  and less/less#2325 like using `sync: true` or `syncImport: true` to no avail.

Indeed, I would like to avoid forking rvc. Question is, though, do I have any other option left?

Is there a way a RactiveJS component’s CSS could be set asynchronously?
